### PR TITLE
Hw spring5 kostevich

### PR DESCRIPTION
**Pull request by gigbat**

### CODE REVIEW

**Review by Likh-Alex — APPROVED** [review]

File: src/main/java/com/dev/cinema/EmailValidatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidatorImpl
+        implements ConstraintValidator<EmailValidator, String> {
+    private static final String regex = "^[A-Za-z0-9+_.-]+@(.+)$";
+
+    @Override
+    public boolean isValid(String field, ConstraintValidatorContext cvc) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/EmailValidatorImpl.java]: `cvc` is not informative

File: src/main/java/com/dev/cinema/PasswordValidator.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    String field();
+
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
+
+    @Target({ ElementType.TYPE })
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        PasswordValidator[] value();
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/PasswordValidator.java]: do we need this?

File: src/main/java/com/dev/cinema/controller/MovieController.java [hunk 3]
```diff
@@ -27,7 +28,7 @@ public MovieController(MovieService movieService, MovieMapper movieMapper) {
     }
 
     @PostMapping
-    public void create(@RequestBody MovieRequestDto movieRequestDto) {
+    public void create(@RequestBody @Valid MovieRequestDto movieRequestDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/MovieController.java:31]: what about MovieSessionController?

File: src/main/java/com/dev/cinema/controller/OrderController.java [hunk 4]
```diff
@@ -37,15 +38,26 @@ public OrderController(UserService userService,
...
+        }
     }
 
     @GetMapping
-    public List<OrderResponseDto> getOrdersHistory(@RequestParam Long userId) {
-        User user = userService.get(userId);
+    public List<OrderResponseDto> getOrdersHistory(Authentication authentication) {
+        Object principal = authentication.getPrincipal();
+        User user = null;
+        if (principal instanceof UserDetails) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/OrderController.java]: let's assume in this case the principal can be safely casted to UserDetails, and remove this check

also, we get get email directly from Authentication
`authentication.getName()`

File: src/main/java/com/dev/cinema/controller/TestController.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.dev.cinema.controller;
+
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/")
+public class TestController {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/TestController.java]: do we need this controller?

File: src/main/java/com/dev/cinema/controller/UserController.java [hunk 6]
```diff
@@ -28,15 +30,19 @@ public UserController(UserService userService,
...
-                -> new RuntimeException("User by email " + email
-                + " wasn't found")));
+        return userMapper.toDto(userByEmail.get());
     }
 
     @PostMapping
-    public void add(@RequestBody UserRequestDto userRequestDto) {
+    public void add(@RequestBody @Valid UserRequestDto userRequestDto) {
         userService.add(userMapper.toEntity(userRequestDto));
     }
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/UserController.java]: this method should not be here - we have `register` for this purpose

File: src/main/java/com/dev/cinema/controller/UserController.java [hunk 7]
```diff
@@ -28,15 +30,19 @@ public UserController(UserService userService,
...
 
     @GetMapping("/by-email")
-    public UserResponseDto findByEmail(@RequestParam String email) {
+    public UserResponseDto findByEmail(Authentication authentication) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/UserController.java]: please see checklist - we should not change this method

File: src/main/java/com/dev/cinema/model/dto/UserRequestDto.java [hunk 8]
```diff
@@ -1,8 +1,26 @@
...
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+
+@PasswordValidator.List({
+        @PasswordValidator(
+                field = "password",
+                fieldMatch = "verifyPassword",
+                message = "Passwords do not match!"
+        )
+})
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/UserRequestDto.java]:
```suggestion
@PasswordValidator
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/EmailValidatorImpl.java (hunk 1 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/EmailValidatorImpl.java]: renamed

File: src/main/java/com/dev/cinema/PasswordValidator.java (hunk 2 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/PasswordValidator.java]: no, but yesterday i thought that it was a need

File: src/main/java/com/dev/cinema/controller/MovieController.java (hunk 3 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/MovieController.java:31]: hmmm, in my code i have it

Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/MovieController.java:31]: image

File: src/main/java/com/dev/cinema/controller/OrderController.java (hunk 4 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/OrderController.java]: done

File: src/main/java/com/dev/cinema/controller/TestController.java (hunk 5 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/TestController.java]: sorry, i forgot to delete that. it was for testing my app)

File: src/main/java/com/dev/cinema/model/dto/UserRequestDto.java (hunk 8 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/model/dto/UserRequestDto.java]: done

File: src/main/java/com/dev/cinema/controller/UserController.java (hunk 6 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/UserController.java]: done

File: src/main/java/com/dev/cinema/controller/UserController.java (hunk 7 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/UserController.java]: done

File: src/main/java/com/dev/cinema/controller/MovieController.java (hunk 3 above)
Comment (KseniiaMakarova) [reply src/main/java/com/dev/cinema/controller/MovieController.java:31]: MovieSessionController and MovieController are different controllers

File: src/main/java/com/dev/cinema/EmailValidatorImpl.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidatorImpl
+        implements ConstraintValidator<EmailValidator, String> {
+    private static final String regex = "^[A-Za-z0-9+_.-]+@(.+)$";
+
+    @Override
+    public boolean isValid(String field, ConstraintValidatorContext validator) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/EmailValidatorImpl.java]:
```suggestion
    public boolean isValid(String field, ConstraintValidatorContext context) {
```

File: src/main/java/com/dev/cinema/PasswordValidator.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import javax.validation.Constraint;
+import javax.validation.Payload;
+
+@Constraint(validatedBy = PasswordValidatorImpl.class)
+@Target({ ElementType.TYPE })
+@Retention(RetentionPolicy.RUNTIME)
+public @interface PasswordValidator {
+    String fieldMatch() default "verifyPassword";
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/PasswordValidator.java]: what is the name of this field in your UserRequestDto?

File: src/main/java/com/dev/cinema/controller/MovieController.java (hunk 3 above)
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/MovieController.java:31]: again, let's also validate dto in MovieSessionController

File: src/main/java/com/dev/cinema/controller/OrderController.java [hunk 11]
```diff
@@ -37,15 +38,22 @@ public OrderController(UserService userService,
     }
 
     @PostMapping("/complete")
-    public void completeOrder(@RequestParam Long userId) {
-        ShoppingCart shoppingCart = shoppingCartService
-                .getByUser(userService.get(userId));
-        orderService.completeOrder(shoppingCart);
+    public void completeOrder(Authentication authentication) {
+        Object principal = authentication.getPrincipal();
+        if (principal instanceof UserDetails) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/OrderController.java]: remove this check

File: src/main/java/com/dev/cinema/controller/OrderController.java [hunk 12]
```diff
@@ -37,15 +38,22 @@ public OrderController(UserService userService,
...
+        }
     }
 
     @GetMapping
-    public List<OrderResponseDto> getOrdersHistory(@RequestParam Long userId) {
-        User user = userService.get(userId);
+    public List<OrderResponseDto> getOrdersHistory(Authentication authentication) {
+        String email = authentication.getName();
+        User user = userService.get(userService
+                .findByEmail(email).get().getId());
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/OrderController.java]: can we simplify this?

File: src/main/java/com/dev/cinema/controller/ShoppingCartController.java [hunk 13]
```diff
@@ -35,15 +37,23 @@ public ShoppingCartController(ShoppingCartService shoppingCartService,
     }
 
     @PostMapping("/movie-sessions")
-    public void addSession(@RequestParam Long userId, Long movieSessionId) {
-        User user = userService.get(userId);
+    public void addSession(Authentication authentication,
+                           @RequestParam Long movieSessionId) {
+        Object principal = authentication.getPrincipal();
+        User user = null;
+        if (principal instanceof UserDetails) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/ShoppingCartController.java]: remove this check

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/EmailValidatorImpl.java (hunk 9 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/EmailValidatorImpl.java]: done

File: src/main/java/com/dev/cinema/PasswordValidator.java (hunk 10 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/PasswordValidator.java]: repeatPassword

File: src/main/java/com/dev/cinema/controller/MovieController.java (hunk 3 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/MovieController.java:31]: done

File: src/main/java/com/dev/cinema/controller/OrderController.java (hunk 11 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/OrderController.java]: done

File: src/main/java/com/dev/cinema/controller/OrderController.java (hunk 12 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/OrderController.java]: done

File: src/main/java/com/dev/cinema/controller/ShoppingCartController.java (hunk 13 above)
Comment (gigbat) [reply src/main/java/com/dev/cinema/controller/ShoppingCartController.java]: done

File: src/main/java/com/dev/cinema/controller/MovieSessionController.java [hunk 14]
```diff
@@ -34,7 +35,7 @@ public MovieSessionController(MovieSessionService movieSessionService,
     }
 
     @PostMapping
-    public void create(@RequestBody MovieSessionRequestDto movieSessionRequestDto) {
+    public void create(@RequestBody @Valid MovieSessionRequestDto movieSessionRequestDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/MovieSessionController.java:38]: should we add validating annotations to MovieSessionRequestDto?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**